core/Rename.cpp

```diff
@@ -83,6 +86,11 @@ namespace olympia
             scoreboards_[rf]->set(bits);
         }
 
+        // initialize freelist
+        for(unsigned int i = 0; i < (int) sparta::Scoreboard::MAX_REGISTERS; i++){
+            freelist[core_types::RegFile::RF_INTEGER].push(i); // INTEGER
+            freelist[core_types::RegFile::RF_FLOAT].push(i); // FLOAT
+        }
```

Comment:
The number of free renames should come from a parameter, not dictated by the size of the scoreboard.  Most hardware designs have a limited number of renames.   

Also, it's not just 1 parameter -- it's a parameter per register file.  For example, there should be one for integer renames, one for float-point renames, etc.

Finally, when you add the parameters, you'll need to break up this `for` loop into each `RegFile` type.  I'd write a `lamda` function that would do this for me.

Reply:
I'm not too familiar with lambda functions, so I wasn't sure of the most elegant way to implement them. The most common example I saw were of one using a switch statement, with a case for each enum, where you would then have a specific parameter for that enum, i.e num_renames_integers:
```
switch(RegFile){
    case(RegFile::RF_INTEGER):
        populate freelist according to num_renames_rf_integers
}
```
but this seemed not too elegant, as you have to define a switch statement. I went ahead and implemented a [map](https://github.com/AaronGChan/riscv-perf-model/blob/fbe7264a97b4aadea92369bdf93a544c3077551e/core/Rename.hpp#L72) ([link](https://github.com/AaronGChan/riscv-perf-model/blob/fbe7264a97b4aadea92369bdf93a544c3077551e/core/Rename.cpp#L93) to usage in setup of Rename)  where the user can define the RegFile type and number of renames for now, but I wanted to get a better idea of the lambda approach.
Also would it be best if users define the number of renames for each register file from the .yaml file? Or would that not make sense, since you would also have to modify the core_types::RegFile enum anyways to add a RegFile?

Reply:
Sure...  this is what I had in mind:
```
    auto setup_map = [this] (core_types::RegFile reg_file, const uint32_t num_renames) {
                         for(uint32_t i = 0; i < num_renames; ++i) {
                             freelist[reg_file].emplace(i);
                         }
                         map_table[reg_file].reset(new std::queue<unsigned int>[num_renames]);
                         reference_counter[reg_file].reset(new num_renames]());
                     };

    setup_map(core_types::RegFile::RF_INTEGER, params->num_integer_renames);
    setup_map(core_types::RegFile::RF_FLOAT,   params->num_float_renames);
    static_assert(N_REGFILES == 2, "New RF type added, but Rename not updated");
```

Reply:
Thanks this made a lot more sense. Added in [163a70b](https://github.com/riscv-software-src/riscv-perf-model/pull/40/commits/163a70b5862658deca0df16b035bbe59dc841c92)